### PR TITLE
Mortar deployer

### PR DESCRIPTION
**Pull request by jnummelin**
This PR integrates [mortar](https://github.com/kontena/mortar/) as a deployer for Skaffold.

Enabling mortar as deployer in skaffold.yaml:
```
deploy:
  mortar:
    name: skaffold-shot
    src: manifests
```

### CODE REVIEW

**Comment by jnummelin** [comment]
The test errors don't seems to be related to my changes, are those tests flaky or did I mess up something?

**Comment by jnummelin** [comment]
I've been trying to track where the test fails. It fails when there's automatic "migration" of the pipeline definition from older version to newer. It fails when processing the yaml tags for un/marshalling. I'm still not sure why adding `MortarDeployer` to the latest schema version would cause that. Maybe for some reason `MortarDeployer` gets initialized un-properly during the migration. I'd really appreciate any tips where to look.

/cc @balopat @dgageot @nkubala @r2d4

**Comment by pre** [comment]
@dgageot Could you help us figure out why this test fails here? The test does succeed when it's run individually. Would be great to get Mortar support in Skaffold :)

**Comment by jnummelin** [comment]
@r2d4 I fixed the linting errors, I had older linter locally... Thanks for pointing me to update it.

**Comment by r2d4** [comment]
Yes the test script unfortunately doesn't enforce the latest version

**Comment by jnummelin** [comment]
@r2d4 Any chance of getting this forward or should I just close it?

File: pkg/skaffold/schema/latest/config.go [hunk 1]
```diff
@@ -148,6 +148,7 @@ type DeployType struct {
 	HelmDeploy      *HelmDeploy      `yaml:"helm,omitempty" yamltags:"oneOf=deploy"`
 	KubectlDeploy   *KubectlDeploy   `yaml:"kubectl,omitempty" yamltags:"oneOf=deploy"`
 	KustomizeDeploy *KustomizeDeploy `yaml:"kustomize,omitempty" yamltags:"oneOf=deploy"`
+	MortarDeploy    *MortarDeploy    `yaml:"mortar,omitempty" yamltags:"oneOf=deploy"`
```
Comment (dgageot) [inline pkg/skaffold/schema/latest/config.go]: Adding new fields to the yaml now requires creating a new version of the schema. I can help you with that once everything else is ok.

File: pkg/skaffold/schema/latest/config.go [hunk 2]
```diff
@@ -220,6 +221,13 @@ type HelmFQNConfig struct {
 type HelmConventionConfig struct {
 }
 
+// MortarDeploy contains the configuration needed for deploying with mortar
+type MortarDeploy struct {
+	Name   string `yaml:"name,omitempty"`
+	Source string `yaml:"src,omitempty"`
+	Config string `yaml:"config,omitempty"`
```
Comment (dgageot) [inline pkg/skaffold/schema/latest/config.go]: Are there default values we could apply to `Source` or `Config`?

Comment (jnummelin) [reply pkg/skaffold/schema/latest/config.go]: Mortar usage dictates always specifying the source, see: https://github.com/kontena/mortar#deploying-k8s-yaml-manifests

Hence I didn't feel like we should have any defaulting for it in Skaffold either.

By default Mortar looks for a file named `shot.yml` or `shot.yaml` as its config file, see https://github.com/kontena/mortar#shot-configuration-file. As the config file is fully optional to use, I donät think we need to default to anything here.

Comment (dgageot) [reply pkg/skaffold/schema/latest/config.go]: 👍

File: pkg/skaffold/deploy/mortar.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	args = append(args, m.MortarDeploy.Source)
+	args = append(args, m.MortarDeploy.Name)
+
+	logrus.Debugf("Firing with args: %v", args)
+	cmd := exec.CommandContext(ctx, "mortar", args...)
+	err := util.RunCmd(cmd)
+	if err != nil {
+		return nil, errors.Wrap(err, "mortar fire")
+	}
+	return nil, nil
```
[the remark this answers is not in the document]
Comment (jnummelin) [reply pkg/skaffold/deploy/mortar.go]: Is the Skaffold labels a strict requirement? Mortar does add some of it's own labels, to find and identify the stack resources.

I see that slightly problematic in this case as typically with Mortar the deployment manifests are actually templates. That would mean that Mortar would first have to parse and evaluate the templates before we can apply the Skaffold specific labels. There is no requirements in Mortar for the templates to be valid yaml BEFORE evaluating.

[the remark this answers is not in the document]
Comment (jnummelin) [reply pkg/skaffold/deploy/mortar.go]: I guess mortar deployer needs to do this in same way as Helm deployer does it.
1. Shoot the manifests
2. Query everything shot (mortar embeds suitable labels for this automatically)
3. re-label everything with skaffold labels

@dgageot WDYT?

**Comment by dgageot** [comment]
@jnummelin Please rebase so that all schema changes are applied to v1beta4

**Comment by jnummelin** [comment]
@dgageot sorry it took a while, I was out-of-office last week. Rebased now. Let me know what else is still needed.

**Comment by jnummelin** [comment]
rebased again. @dgageot PTAL and ping me if anything is still needed

**Comment by balopat** [comment]
Hi @jnummelin - I would like to apologize for holding out without a clear response and giving mixed signals this long. We decided not to add and maintain any other deployers as part of the skaffold codebase, and until we see more demand, I would put this idea on hold and close this issue and the PR.

**Comment by matti** [comment]
@balopat please consider this again, mortar has by far the best balance of features (template language, variables, overlays) vs helm and kustomize.

**Pull request closed without merging**